File: server/game-engine/src/main/java/com/deepdame/engine/core/model/GameState.java
package com.deepdame.engine.core.model;

import org.springframework.data.annotation.PersistenceCreator;

import java.util.UUID;

public class GameState {

    private UUID id;
    private Board board;
    private PieceType currentTurn;
    private boolean isGameOver;
    private PieceType winner;

    public GameState(UUID id){
        this.id = id;
        this.board = new Board();
        this.currentTurn = PieceType.BLACK;
        this.isGameOver = false;
        this.winner = null;
    }

    @PersistenceCreator
    public GameState(UUID id, Board board, PieceType currentTurn, boolean isGameOver, PieceType winner) {
        this.id = id;
        this.board = board;
        this.currentTurn = currentTurn;
        this.isGameOver = isGameOver;
        this.winner = winner;
    }

    public GameState(){}

    public void switchTurn(){
        if (isGameOver) return;
        this.currentTurn = (this.currentTurn == PieceType.BLACK) ? PieceType.WHITE : PieceType.BLACK;
    }

    public void finishGame(PieceType winner){
        this.isGameOver = true;
        this.winner = winner;
    }

    public UUID getId() { return id; }
    public Board getBoard() { return board; }
    public PieceType getCurrentTurn() { return currentTurn; }
    public boolean isGameOver() { return isGameOver; }
    public PieceType getWinner() { return winner; }
}

File: client/deepdame/lib/game-engine/model/game_state.dart
import 'board.dart';
import 'piece_type.dart';

class GameState {
  
  final String id;
  
  final Board board;
  
  PieceType _currentTurn;
  bool _isGameOver;
  PieceType? _winner;

  GameState.newGame(this.id)
      : board = Board(),
        _currentTurn = PieceType.black,
        _isGameOver = false,
        _winner = null;

  GameState.load({
    required this.id,
    required this.board,
    required PieceType currentTurn,
    required bool isGameOver,
    PieceType? winner,
  })  : _currentTurn = currentTurn,
        _isGameOver = isGameOver,
        _winner = winner;

  void switchTurn() {
    if (_isGameOver) return;
    _currentTurn = (_currentTurn == PieceType.black) 
        ? PieceType.white 
        : PieceType.black;
  }

  void finishGame(PieceType winner) {
    _isGameOver = true;
    _winner = winner;
  }

  PieceType get currentTurn => _currentTurn;
  bool get isGameOver => _isGameOver;
  PieceType? get winner => _winner;
}